README.md
express-sslify
==============

This simple module enforces HTTPS connections on any incoming GET and HEAD requests. In case of a non-encrypted HTTP request, express-sslify automatically redirects to an HTTPS address using a 301 permanent redirect. Any other type of request (e.g., POST) will fail with a 403 error message.

express-sslify also works behind reverse proxies (load balancers) such as those used by Heroku or nodejitsu. In those cases, however, the `trustProtoHeader` parameter has to be set (see below)

### Usage

First, install the module:

`$ npm install express-sslify --save`

Afterwards, require the module and *use* the `HTTPS()` method:
```javascript
var express = require('express');
var http = require('http');
var enforce = require('express-sslify');

var app = express();

// Use enforce.HTTPS({ trustProtoHeader: true }) in case you are behind
// a load balancer (e.g. Heroku). See further comments below
app.use(enforce.HTTPS());

http.createServer(app).listen(app.get('port'), function() {
	console.log('Express server listening on port ' + app.get('port'));
});
```

### Reverse Proxies (Heroku, nodejitsu and others)

Heroku, nodejitsu and other hosters often use reverse proxies which offer SSL endpoints but then forward unencrypted HTTP traffic to the website. This makes it difficult to detect if the original request was indeed via HTTPS. Luckily, most reverse proxies set the `x-forwarded-proto` header flag with the original request scheme. express-sslify is ready for such scenarios, but you have to specifically request the evaluation of this flag:

```javascript
app.use(enforce.HTTPS({ trustProtoHeader: true }))
```

Please do *not* set this flag if you are not behind a proxy that is setting this flag. HTTP headers can be easily spoofed outside of environments that are actively setting/removing the header.

### Azure support

Azure has a slightly different way of signaling encrypted connections. To tell express-sslify to look out for Azure's x-arr-ssl header do the following:

```javascript
app.use(enforce.HTTPS({ trustAzureHeader: true }))
```

Please do *not* set this flag if you are not behind an Azure proxy as this flag can be easily spoofed outside of an Azure environment.

### X-Forwarded-Host header support

If your reverse proxy sends the original host using the `X-Forwarded-Host` header and you need to use that instead of the `Host` header for the redirect, use the `trustXForwardedHostHeader` flag:

```javascript
app.use(enforce.HTTPS({ trustXForwardedHostHeader: true }))
```

## Tests
Download the whole repository and call:
```shell
$ npm install; npm test
```

### Credits and License
express-sslify is licensed under the MIT license. If you'd like to be informed about new projects follow  [@TheSumOfAll](http://twitter.com/TheSumOfAll/).

Copyright (c) 2013-2016 Florian Heinemann


CHANGELOG.md
# 1.2.0 - Dec 22 2016
- ADD Support for 'x-forwarded-host' headers to indicate to which host requests should be redirected

# 1.1.0 - Jul 02 2016
- ADD Support for 'x-forwarded-proto' headers that were appended by several proxies in the format 'https, http'
- UPDATE Tests for Azure HTTPS to have more realistic content for the x-arr-ssl flag
- UPDATE JSDoc and Readme

# 1.0.1 - Oct 25 2015
- ADD redirects for HEAD requests

# 1.0.0 - Oct 24 2015
- BREAKING change of arguments to named arg. - check README
- UPDATE documentation

# 0.1.2 - Jul 06 2015
- UPDATE dependencies
- BUMP package version

# 0.1.1 - Nov 15 2014
- ADDED Azure proxy support
- ADDED Testing


test/express-sslify.test.js
'use strict';

var expect = require('chai').expect;
var express = require('express');
var request = require('supertest');
var enforce = require('../index.js');

describe('express-sslify', function() {
	describe('HTTPS not enforced', function() {

		var app = express();

		app.get('/non-ssl',
			function(req, res){
				res.status(200).send('ok');
		});

		app.head('/non-ssl-head',
			function(req, res){
				res.status(200).send();
		});

		app.post('/non-ssl-post',
			function(req, res){
				res.status(200).send('ok');
		});

		var agent = request.agent(app);

		it('should accept non-ssl requests', function (done) {
			agent
				.get('/non-ssl')
				.expect(200, 'ok', done);
		})

		it('should accept non-ssl HEAD requests', function (done) {
			agent
				.head('/non-ssl-head')
				.expect(200, done);
		})

		it('should accept non-ssl POST requests', function (done) {
			agent
				.post('/non-ssl-post')
				.expect(200, 'ok', done);
		})
	})

	describe('HTTPS enforced', function() {

		var app = express();

		app.use(enforce.HTTPS());

		app.get('/ssl',
			function(req, res){
				res.status(200).send('ok');
		});

		app.head('/ssl-head',
			function(req, res){
				res.status(200).send();
		});

		app.post('/ssl-post',
			function(req, res){
				res.status(200).send('ok');
		});

		var agent = request.agent(app);

		it('should redirect non-SSL GET requests to HTTPS', function (done) {
			agent
				.get('/ssl')
				.expect(301)
				.expect('location', new RegExp('^https://[\\S]*/ssl$'), done);
		})

		it('should redirect non-SSL HEAD requests to HTTPS', function (done) {
			agent
				.head('/ssl-head')
				.expect(301)
				.expect('location', new RegExp('^https://[\\S]*/ssl-head$'), done);
		})

		it('should send error for non-SSL POST requests', function (done) {
			agent
				.post('/non-ssl-post')
				.expect(403, done);
		})
	})

	describe('Heroku-style proxy SSL flag', function() {

		var proxyTests = function(method) {

			var app = express();

			app[method]('/ssl', enforce.HTTPS(),
				function(req, res){
					res.status(200).send();
			});

			app[method]('/ssl-behind-proxy', enforce.HTTPS({ trustProtoHeader: true }),
				function(req, res){
					res.status(200).send();
			});

			var agent = request.agent(app);

			it('should ignore x-forwarded-proto if not activated (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl')
	      			.set('x-forwarded-proto', 'https')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl$'), done);
			})

			it('should accept request if flag set and activated (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-proxy')
	      			.set('x-forwarded-proto', 'https')
					.expect(200, done);
			})

			it('should accept request if flag set and activated (' + method.toUpperCase() + ') with comma/space separated list', function (done) {
				agent
					[method]('/ssl-behind-proxy')
					.set('x-forwarded-proto', 'https, http')
					.expect(200, done);
			})

			it('should accept request if flag set and activated (' + method.toUpperCase() + ') with comma separated list', function (done) {
				agent
					[method]('/ssl-behind-proxy')
					.set('x-forwarded-proto', 'https,http')
					.expect(200, done);
			})

			it('should redirect if activated but flag not set with https (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-proxy')
					.set('x-forwarded-proto', '')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-proxy$'), done);
			})

			it('should redirect if activated but flag only set with HTTP (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-proxy')
					.set('x-forwarded-proto', 'http')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-proxy$'), done);
			})

			it('should redirect if activated but header indicates that first hop was not HTTPS (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-proxy')
					.set('x-forwarded-proto', 'http, https')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-proxy$'), done);
			})

			it('should redirect if activated but flag not set (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-proxy')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-proxy$'), done);
			})

			it('should redirect if activated but wrong flag set (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-proxy')
	      			.set('x-arr-ssl', 'https')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-proxy$'), done);
			})
		}

		// Test GET requests
		proxyTests('get');

		// Test HEAD requests
		proxyTests('head');
	})

	describe('Azure-style proxy SSL flag', function() {

		var proxyTests = function(method) {

			var app = express();

			var xArrSslContent = '2048|128|DC=com, DC=microsoft, DC=corp, DC=redmond, CN=MSIT Machine Auth CA 2|C=US, S=WA, L=Redmond, O=Microsoft, OU=OrganizationName, CN=*.azurewebsites.net';

			app[method]('/ssl', enforce.HTTPS(),
				function(req, res){
					res.status(200).send();
			});

			app[method]('/ssl-behind-azure', enforce.HTTPS({trustAzureHeader: true}),
				function(req, res){
					res.status(200).send();
			});

			var agent = request.agent(app);

			it('should ignore x-arr-ssl if not activated (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl')
	      			.set('x-arr-ssl', xArrSslContent)
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl$'), done);
			})

			it('should accept request if flag set and activated (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-azure')
	      			.set('x-arr-ssl', xArrSslContent)
					.expect(200, done);
			})

			it('should redirect if activated but flag not set (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-azure')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-azure$'), done);
			})

			it('should redirect if activated but wrong flag set (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-behind-azure')
	      			.set('x-forwarded-proto', 'https')
					.expect(301)
					.expect('location', new RegExp('^https://[\\S]*/ssl-behind-azure$'), done);
			})
		}

		// Test GET requests
		proxyTests('get');

		// Test HEAD requests
		proxyTests('head');
	})

	describe('X-Forwarded-Host redirects', function() {

		var proxyTests = function(method) {

			var app = express();

			var xArrSslContent = '2048|128|DC=com, DC=microsoft, DC=corp, DC=redmond, CN=MSIT Machine Auth CA 2|C=US, S=WA, L=Redmond, O=Microsoft, OU=OrganizationName, CN=*.azurewebsites.net';

			app[method]('/ssl', enforce.HTTPS(),
				function(req, res){
					res.status(200).send();
			});

			app[method]('/ssl-with-redirect-trusted', enforce.HTTPS({trustXForwardedHostHeader: true}),
				function(req, res){
					res.status(200).send();
			});

			var agent = request.agent(app);

			it('should ignore x-forwarded-host if not activated (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl')
	      			.set('x-forwarded-host', 'malicious')
					.expect(301)
					.expect(function(res) {
						if(res.header.location.indexOf('malicious') != -1)
							throw new Error('should not redirect')
					})
					.expect('location', new RegExp('^https://[\\S]*/ssl$'), done);
			})

			it('should ignore x-forwarded-host if not set (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-with-redirect-trusted')
					.expect(301)
					.expect(function(res) {
						if(res.header.location.indexOf('localhost') != -1 && res.header.location.indexOf('127.0.0.1') != -1)
							throw new Error('should not redirect')
					})
					.expect('location', new RegExp('^https://[\\S]*/ssl-with-redirect-trusted$'), done);
			})

			it('should redirect if x-forwarded-host and flag is set (' + method.toUpperCase() + ')', function (done) {
				agent
					[method]('/ssl-with-redirect-trusted')
	      			.set('x-forwarded-host', 'newhost123')
					.expect(301)
					.expect(function(res) {
						if(res.header.location.indexOf('newhost123') === -1)
							throw new Error('should redirect')
					})
					.expect('location', new RegExp('^https://[\\S]*/ssl-with-redirect-trusted$'), done);
			})
		}

		// Test GET requests
		proxyTests('get');

		// Test HEAD requests
		proxyTests('head');
	})

	describe('Pre-1.0.0-style arguments', function() {

		var app = express();

		app.get('/ssl', enforce.HTTPS(true),
			function(req, res){
				res.status(200).send('ok');
		});

		var agent = request.agent(app);

		it('should crash', function (done) {
			agent
				.get('/ssl')
      			.set('x-forwarded-proto', 'https')
				.expect(500, done);
		})
	})
})


index.js
"use strict";

var defaults = {
	trustProtoHeader: false,
	trustAzureHeader: false,
	trustXForwardedHostHeader: false
};

/**
 * Apply options
 *
 * @param {Hash} [options]
 * @return {Hash}
 * @api private
 */
function applyOptions(options) {
	var settings = {};
	options = options || {};

	for (var option in defaults) {
		settings[option] = options[option] || defaults[option];
	}
	return settings;
}

/**
 * enforceHTTPS
 *
 * @param {Hash} [options]
 * @param {Boolean} [options[trustProtoHeader]=false] - Set to true if the x-forwarded-proto HTTP header should be trusted (e.g. for typical reverse proxy configurations)
 * @param {Boolean} [options[trustAzureHeader]=false] - Set to true if Azure's x-arr-ssl HTTP header should be trusted (only use in Azure environments)
 * @param {Boolean} [options[trustXForwardedHostHeader]=false] - Set to true if the x-forwarded-host HTTP header should be trusted
 * @api public
 */
var enforceHTTPS = function(options) {
	return function(req, res, next) {
		// Crash on pre-1.0.0-style arguments
		if(typeof options === 'boolean') {
			return next("express-sslify has changed the way how arguments are treated. Please check the readme.");
		}

		options = applyOptions(options);

		// First, check if directly requested via https
		var isHttps = req.secure;

		// Second, if the request headers can be trusted (e.g. because they are send
		// by a proxy), check if x-forward-proto is set to https
		if(!isHttps && options.trustProtoHeader) {
			isHttps = ((req.headers["x-forwarded-proto"] || '').substring(0,5) === 'https');
		}

		// Third, if trustAzureHeader is set, check for Azure's headers
		// indicating a SSL connection
		if(!isHttps && options.trustAzureHeader && req.headers["x-arr-ssl"]) {
			isHttps = true;
		}

		if(isHttps) {
			next();
		} else {
			// Only redirect GET methods
			if(req.method === "GET" || req.method === 'HEAD') {
				var host = options.trustXForwardedHostHeader ? (req.headers['x-forwarded-host'] || req.headers.host) : req.headers.host;
				res.redirect(301, "https://" + host + req.originalUrl);
			} else {
				res.status(403).send("Please use HTTPS when submitting data to this server.");
			}
		}
	};
};

exports.HTTPS = enforceHTTPS;
